### PR TITLE
Fix/26040 null routerlink overrides href

### LINKED ISSUE #26040: [routerLink]="null" should not affect the href attribute
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ ] Performance issue
[x] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
[ ] Other... Please describe:
</code></pre>

## Current behavior
the following template: `<a href="my/custom/url" [routerLink]="null">link</a>`
is generated to: `<a href="current/page/url">link</a>`

## Expected behavior
the following template: `<a href="my/custom/url" [routerLink]="null">link</a>`
should be generated to: `<a href="my/custom/url">link</a>`

## Minimal reproduction of the problem with instructions
https://stackblitz.com/edit/angular-gitter-63xra2?file=app%2Fapp.component.html

## What is the motivation / use case for changing the behavior?
In case you have an array of links to handle, and the array includes links with 'routerLink routes' and 'normal (pl absolute-) URLs'. Then it's much easier to handle them in a template (see: reproduction example)

## Environment

<pre><code>
Angular version: 6.1.7

### PR DESCRIPTION
**Pull request by maximelafarie**
Fix a bug where a null value passed to routerLink directive overrides the existing href.

fixes #26040

Previous PR: https://github.com/angular/angular/pull/27347 (but I guess it has been left aside).

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature

## What is the current behavior?
Description and full Stackblitz example are in the issue below: 

Issue Number: #26040 

## What is the new behavior?
If a `null` value is passed to `routerLink` directive and if an `href` attribute is defined and not empty, `href` will remain unchanged. 

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by maximelafarie** [comment]
Maybe CI need to be restarted? The failure isn't related to the PR itself. 🤓 

(@jasonaden still nothing? Maybe for the v9 release it may be left aside... but it's here since November 2018! 😇)

File: packages/router/src/router_state.ts [hunk 1]
```diff
@@ -321,7 +321,7 @@ export class ActivatedRouteSnapshot {
   }
 
   toString(): string {
-    const url = this.url.map(segment => segment.toString()).join('/');
+    const url = this.url ? this.url.map(segment => segment.toString()).join('/') : '';
```
Comment (atscott) [inline packages/router/src/router_state.ts]: Is this change relevant to the current PR? If not, please revert this change and create a new PR/issue for why it's needed. Also, I tried reverting the change locally and all the tests in router still pass so you'll also need a test to demonstrate why this change is necessary (in the new PR or this one if it's needed for this PR).

File: packages/router/src/directives/router_link.ts [hunk 2]
```diff
@@ -215,7 +215,7 @@ export class RouterLinkWithHref implements OnChanges, OnDestroy {
 
   // the url displayed on the anchor element.
   // TODO(issue/24571): remove '!'.
-  @HostBinding() href !: string;
+  @HostBinding('attr.href') @Input() href !: string;
```
Comment (atscott) [inline packages/router/src/directives/router_link.ts]: What's the reason for the 'attr.href' change? Reverting it does not appear to cause any tests to fail.

**Comment by atscott** [comment]
I rebased your PR and squashed your commits into one. Please address any CI failures (if there are any when it finishes)

**Comment by maximelafarie** [comment]
Thank you @atscott there's a long time I made this PR I'll take a look at it again and do the right stuff were needed 👍

**Comment by atscott** [comment]
Closing this in favor of bug report #13980. `routerLink` inherently modifies the `href` attribute. Setting it to `null` should clear the `href` instead of leaving it alone. If a `routerLink` that does not affect the href attribute is desired, I would recommend copying the `RouterLinkWithHref` code and creating your own directive for this special case.
